fastlane_bot/tools/cpc.py

```diff
@@ -1555,6 +1700,46 @@ def isbigger(self, small, big):
             return big > self.EPS
         return big / small > 1 + self.EPS
 
+    def plot(self, xmin=None, xmax=None, steps=None, *, xvals=None, func=None, show=False, title=None, xlabel=None, ylabel=None, grid=True, **params):
+        """
+        plots the curve associated with this pool
+        
+        :xmin, xmax, steps:    x range (args for np.linspace)
+        :xvals:                x values (alternative to xmin, xmax, steps)
+        :func:                 function to plot (default: dyfrpmdx_f)
+        :show:                 if True, call plt.show()
+        :title:                plot title
+        :xlabel, ylabel:       axis labels
+        :grid:                 if True [False], [do not] show grid; None: ignore
+        :params:               additional kwargs passed to plt.plot
+        """
+        if xvals is None:
+            assert not xmin is None, "xmin must not be None if xv is None"
+            assert not xmax is None, "xmin must not be None if xv is None"
+            x_v = np.linspace(xmin, xmax, steps) if steps else np.linspace(xmin, xmax)
+        else:
+            assert xmin is None, "xmin must be None if xv is not None"
+            assert xmax is None, "xmax must be None if xv is not None"
+            assert steps is None, "steps must be None if xv is not None"
+            x_v = xvals
+            
+        xlabel = xlabel or (f"dx [{self.tknx}]" if not func else "x")
+        ylabel = ylabel or (f"dy [{self.tkny}]" if not func else "y")
```

Comment:
Using `or` between two values, one of which is a non-empty string.
Isn't that always going to produce `True`?

Reply:
1. I cannot see the equivalence here
2. I cannot see how it's related to what I wrote

Reply:
the expression works as follows

xlabel given -> xlabel
not func -> f"dx [{self.tknx}]" 
else -> x

Reply:
<img width="198" alt="Screenshot 2024-03-21 at 13 17 03" src="https://github.com/bancorprotocol/fastlane-bot/assets/109073706/43893f7e-5352-4d08-9fd0-be66290d543f">

Reply:
I am not familiar with `or` being anything other than a boolean operator (i.e., the expression `x or y` is always evaluated to either True or False as far as I'm aware of).

And even if this is not the case, then for the purpose of readability, I'd restrict my usage of this operator only as a boolean one.

As far as I understand `xlabel = xlabel or (f"dx [{self.tknx}]" if not func else "x")`, it is equivalent to:
```
if not xlabel:
    xlabel = "x" if func else f"dx [{self.tknx}]"
```
If this analysis is correct, then I'd rewrite that statement as such (again, for the purpose of readability).

Reply:
Barak, respectfully: we are writing code for people who are familiar with Python; if someone stumbles upon this expression which uses a well-known Python construct and wonders how it works then they should probably take some time to look up the finer details of the language and not assume it is JavaScript or C++

Reply:
You asked for a code review, I'm giving you one; feel free to do whatever you want with it :)

Reply:
siding with @sklbancor here - as it's pretty standard for OR expression to serve as means to return the first truthy value. Even in JS for example:
<img width="265" alt="image" src="https://github.com/bancorprotocol/fastlane-bot/assets/672016/e9019e63-2ce5-4d68-9178-783da18148e1">

Reply:
But this is a param passed from outside - unless you're suggesting using this option in the default params?

Reply:
regarding efficiency -- keep in mind that everything in Python is an object, so essentially that is a simple 64-bit write, or whatever the length of an address field is. In an interpreted language this does not take any meaningful time, relatively speaking.

Reply:
BTW, you may by the least want to change your "if not func" logic to "if func", which - again - makes it slightly easier to read. In either case, this thread has spanned across 998 more comments than what it was initially designated for, so feel free to resolve the comment and fix the code whichever way you see fits.

Reply:
Fun fact:

`x = x or None`  is actually (slightly) faster (!!!) than the version with if

<img width="434" alt="Screenshot 2024-03-21 at 14 35 22" src="https://github.com/bancorprotocol/fastlane-bot/assets/109073706/3d480bf4-a41b-449e-86a2-dd68f4a35e4b">